src/libstd/sys/unix/fs.rs

```diff
@@ -132,6 +139,32 @@ impl FromInner<raw::mode_t> for FilePermissions {
 impl Iterator for ReadDir {
     type Item = io::Result<DirEntry>;
 
+    #[cfg(target_os = "sunos")]
```

Comment:
> it may be best to call readdir_r anyway because we already have to store the path locally

Well, in this case I guess it's required to pre-allocate memory to store the path - the specific instructions from the [man page](https://illumos.org/man/3C/readdir) are:

       The caller must allocate storage pointed to by entry to be large enough
       for a dirent structure with an array of char d_name member containing at
       least NAME_MAX (that is, pathconf(directory, _PC_NAME_MAX)) plus one
       elements. (_PC_NAME_MAX is defined in <unistd.h>.)

I.e. that'd also require a separate handling too, as it seems.

I'm not quite sure about what would work best here, TBH, so for now I've added some comments describing the current implementation. Please let me know if it requires polishing or a better implementation!